Updated Panstarrs ingestion

drmegannewsome opened the pull request:
Removed outdated links and downloading protocol for Panstarrs reference image ingestion, replacing with new links that auto-download without user interaction. The new setup still tiles together multiple images around the central RA and Dec to cover the frame.

abostroem commented:
@drmegannewsome thanks for the PR - updates are always great. To facilitate the review of this PR can you tell us how you tested the code changes in the PR?

drmegannewsome commented:
Previously the use of `-s ingestps1` would create a .txt file, asking the user to submit the file to an outdated site to manually download PS1 images for templates. This update allows the user to run, e.g., 
`lscloop.py -n AT2020mot -e 20201017 -d 291 -s ingestps1`
which will automatically download the template images, tile and stack them, and ingest them into the database.

griffin-h commented on trunk/src/lsc/externaldata.py:
```diff
@@ -294,6 +295,7 @@ def sdss_swarp(imglist,_telescope='spectral',_ra='',_dec='',output='', objname='
 
        try:
           _mjd = MJDnow(datetime.datetime(int(str(_dayobs)[0:4]),int(str(_dayobs)[4:6]),int(str(_dayobs)[6:8])))
+          _ut = jd2date(_mjd+2400000.5).strftime('%H%M%S')
```
Am I right that this is always going to return 00:00:00? The MJD in the previous line is constructed from a datetime of the form YYYYMMDD, so I don't think there is any time information.

griffin-h replied:
Also does this get used anywhere? I think this is inside the SDSS section.

griffin-h replied:
Megan says:
> To address the comment about _ut time, the line _ut = jd2date(_mjd+2400000.5).strftime('%H%M%S') does not just return 00:00:00. The database saves the PS1 templates with non-zero UT's. We also made this change because we receive an error when an explicit line defining _ut wasn't defined; thus we added it to the SDSS part as well for completeness' sake.

There's a couple things here:

First of all, I am asking about `_ut` always equaling 00:00:00 for the *SDSS* case, not PS1. I'm pretty sure this will make all our SDSS templates have UTSTART=00:00:00.

Second, the error you're talking about where `_ut` is not defined is because of the code you added below where you introduce UT and UTSTART has header keywords:
```
    hd['UTSTART']  = (_ut,        'universal time')
    hd['UT']       = (_ut,        'universal time')
```
This seems not necessary because we never had it for SDSS before. Would everything run if you just removed all this UT stuff? If the choice is between including UT for PS1 but setting it to the wrong value for PS1, or not including it at all, I think I'd rather not include it.

griffin-h replied:
I moved this line outside the conditional so that the UT is calculated from the MJD for both SDSS and PS1. I also removed the keyword `UT` from the header because `UTSTART` is the standard keyword for LCO data.

griffin-h commented on trunk/src/lsc/externaldata.py:
```diff
@@ -303,21 +305,23 @@ def sdss_swarp(imglist,_telescope='spectral',_ra='',_dec='',output='', objname='
+           hdr = fits.open(img[0])
+           if os.path.isfile(re.sub('unconv','unconv_1',img[0])):
+              os.system('rm '+re.sub('unconv','unconv_1',img[0]))
+           #import pdb; pdb.set_trace()
```
Let's make sure to clean up all these commented lines from debugging.

griffin-h commented on trunk/src/lsc/externaldata.py:
```diff
@@ -337,10 +341,9 @@ def sdss_swarp(imglist,_telescope='spectral',_ra='',_dec='',output='', objname='
-    print _dateobs, _dayobs, _mjd, _filter
+       _dateobs = jd2date(_mjd+2400000.5).strftime('%Y-%m-%d')
+       
+    #print _dateobs, _dayobs, _ut, _mjd, _filter
```
Same comment as above about deleting commented-out lines.

griffin-h commented on trunk/src/lsc/externaldata.py:
```diff
@@ -420,17 +424,20 @@ def sdss_swarp(imglist,_telescope='spectral',_ra='',_dec='',output='', objname='
     hd['L1FWHM']   = (9999,       'FHWM (arcsec) - computed with sextractor')
     hd['WCSERR']   = (0,          'error status of WCS fit. 0 for no error')
-    hd['MJD-OBS']  = (_mjd,       'MJD')
+    hd['MJD']      = (_mjd,       'MJD')
+    hd['MJD_OBS']  = (_mjd,       'MJD')
```
Why did you change these keywords? The standard keyword for LCO images is MJD-OBS.

griffin-h replied:
Megan says:
> The header keywords "MJD"/"MJD_OBS" needed to be changed to those values because we were getting a keyword error when using "MJD-OBS" otherwise. This has been confirmed by running with the old keyword header and getting an error of keyword not found.

The issue I'm worried about is that now you have *removed* `MJD-OBS` for the SDSS images. Have you tested that this branch of the code does not break SDSS ingestion?

griffin-h replied:
I confirmed that `MJD-OBS` is the keyword recognized by the pipeline, so I reverted this change.

griffin-h commented on trunk/src/lsc/externaldata.py:
```diff
@@ -481,6 +488,65 @@ def northupeastleft(filename='', data=None, header=None):
+        url = []
+        for filename in table['filename']:
+            url.append(urlbase+filename)
+    #fitsurl = geturl(ra, dec, size=size, filters="i", format="fits")
+    #fh = fits.open(url[0])
+    #fim = fh[0].data
```
Same comment as above about deleting commented-out lines.

griffin-h commented on trunk/src/lsc/externaldata.py:
```diff
@@ -514,8 +580,22 @@ def sloanimage(img,survey='sloan',frames=[], show=False, force=False):
-      frames =  downloadsdss(_ra, _dec, _band, _radius, force)
+      frames = downloadsdss(_ra, _dec, _band, _radius, force)
    elif survey == 'ps1':
+      delta= 0.15
+      DR = delta/np.cos(_dec*np.pi/180)
+      DD = delta
```
Can you add a comment about what's going on here?

griffin-h replied:
Megan says:
> The line DR = delta/np.cos(_dec*np.pi/180) was actually already in the original PS1 ingestion version (original line 521; I just put it in a new place), so this is not a change. This line selects images at equal distances from the RA/dec of the source.

I'm going to add this comment in the code.

griffin-h commented on trunk/src/lsc/externaldata.py:
```diff
@@ -556,8 +636,9 @@ def sloanimage(img,survey='sloan',frames=[], show=False, force=False):
                     frames2.append(img)
 
           frames = frames2
+   '''
```
Same comment as above about deleting commented-out lines.

abostroem commented:
@drmegannewsome here's a PR you can use as an example for formatting and what information to include: #118

drmegannewsome commented:
The latest push addresses issues of: outdated/excessive commentary lines and patchy image tiling. This version successfully downloads and tiles the images from PS1 to cover the same frame as observed by LCO without missing sections. I have had no timeout issues on multiple tests. I have been encouraged to merge this for GW followup purposes for rapid, real-time subtractions of galaxy candidates.

drmegannewsome commented:
The header keywords `"MJD"/"MJD_OBS" `needed to be changed to those values because we were getting a keyword error when using "MJD-OBS" otherwise. This has been confirmed by running with the old keyword header and getting an error of keyword not found.

To address the comment about _ut time, the line `_ut = jd2date(_mjd+2400000.5).strftime('%H%M%S')` does not just return 00:00:00. The database saves the PS1 templates with non-zero UT's. We also made this change because we receive an error when an explicit line defining _ut wasn't defined; thus we added it to the SDSS part as well for completeness' sake.

The line `DR = delta/np.cos(_dec*np.pi/180)` was actually already in the original PS1 ingestion version (original line 521; I just put it in a new place), so this is not a change. This line selects images at equal distances from the RA/dec of the source.

griffin-h commented:
> The header keywords `"MJD"/"MJD_OBS" `needed to be changed to those values because we were getting a keyword error when using "MJD-OBS" otherwise. This has been confirmed by running with the old keyword header and getting an error of keyword not found.
> 
> To address the comment about _ut time, the line `_ut = jd2date(_mjd+2400000.5).strftime('%H%M%S')` does not just return 00:00:00. The database saves the PS1 templates with non-zero UT's. We also made this change because we receive an error when an explicit line defining _ut wasn't defined; thus we added it to the SDSS part as well for completeness' sake.
> 
> The line `DR = delta/np.cos(_dec*np.pi/180)` was actually already in the original PS1 ingestion version (original line 521; I just put it in a new place), so this is not a change. This line selects images at equal distances from the RA/dec of the source.

I replied to these in the threads above.

griffin-h commented:
In Slack, Megan mentioned that she had already ingested SDSS and PS1 templates for AT 2022upj. I looked at the headers of those images:
```
(pipeline) griffin@supernova:/supernova/data/extdata/20120529$ gethead sinistro_PS1_20120529_gp_AT2022upj.fits mjd mjd-obs mjd_obs ut utstart
sinistro_PS1_20120529_gp_AT2022upj.fits 56076.8954852977 55909.61602194 56076.8954852977 212929 212929
```
```
(pipeline) griffin@supernova:/supernova/data/extdata/20060917$ gethead sinistro_SDSS_20060917_gp_AT2022upj.fits mjd mjd-obs mjd_obs ut utstart
sinistro_SDSS_20060917_gp_AT2022upj.fits 53995.0 53995.00000000 53995.0 000000 000000
```
I notice three problems:
1. `MJD-OBS` (the standard LCO keyword) is now different than `MJD` and `MJD_OBS`. I suggest we just delete the latter two and set the standard one correctly.
2. The times (`UT` and `UTSTART`) are stored without colons in the header (212929 instead of 21:29:29).
3. The times are not set correctly for the SDSS image (000000).

griffin-h commented:
I tested this on SN 2022xxf:
```
lscloop.py -n 'SN 2022xxf' -e 20230626 -d 33 -s ingestsloan
lscloop.py -n 'SN 2022xxf' -e 20230626 -d 33 -s ingestps1
```

This shows that the header keywords are set correctly. (SDSS does not come with times.)
```
(pipeline) griffin@supernova:~$ gethead /supernova/data/extdata/20021211/*SN2022xxf.fits /supernova/data/extdata/20111113/*SN2022xxf.fits mjd-obs date-obs
/supernova/data/extdata/20021211/sinistro_SDSS_20021211_gp_SN2022xxf.fits 52619.0 2002-12-11
/supernova/data/extdata/20111113/sinistro_PS1_20111113_gp_SN2022xxf.fits  55878.3518878931 2011-11-13T08:26:43.113949
```
The metadata also looks good on SNEx, so I'm going to go ahead and merge.
